router/internal/persistedoperation/cdn/client.go

```diff
@@ -0,0 +1,171 @@
+	organizationID string
+	httpClient     *http.Client
+	logger         *zap.Logger
+	tracer         trace.Tracer
+}
+
+func (cdn *client) PersistedOperation(ctx context.Context, clientName string, sha256Hash string, attributes []attribute.KeyValue) ([]byte, error) {
+
+	ctx, span := cdn.tracer.Start(ctx, "Load Persisted Operation",
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(attributes...),
+	)
+	defer span.End()
+
+	content, err := cdn.persistedOperation(ctx, clientName, sha256Hash, attributes)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return nil, err
+	}
+
+	return content, nil
+}
+
+func (cdn *client) persistedOperation(ctx context.Context, clientName string, sha256Hash string, attributes []attribute.KeyValue) ([]byte, error) {
+
+	span := trace.SpanFromContext(ctx)
+
+	operationPath := fmt.Sprintf("/%s/%s/operations/%s/%s.json",
+		cdn.organizationID,
+		cdn.federatedGraphID,
+		url.PathEscape(clientName),
+		url.PathEscape(sha256Hash))
+	operationURL := cdn.cdnURL.ResolveReference(&url.URL{Path: operationPath})
+
+	req, err := http.NewRequestWithContext(ctx, "GET", operationURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	span.SetAttributes(
+		semconv.HTTPURL(req.URL.String()),
+		semconv.HTTPMethod(http.MethodGet),
+		semconv12.HTTPHostKey.String(req.Host),
+	)
+
+	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+	req.Header.Add("Authorization", "Bearer "+cdn.authenticationToken)
+	req.Header.Set("Accept-Encoding", "gzip")
+
+	resp, err := cdn.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	span.SetAttributes(semconv.HTTPStatusCode(resp.StatusCode))
+
+	if resp.StatusCode != http.StatusOK {
+		span.SetStatus(codes.Error, fmt.Sprintf("unexpected status code when loading persisted operation, statusCode: %d", resp.StatusCode))
+
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, &persistedoperation.PersistentOperationNotFoundError{
+				ClientName: clientName,
+				Sha256Hash: sha256Hash,
+			}
+		}
+		if resp.StatusCode == http.StatusUnauthorized {
+			return nil, errors.New("could not authenticate against CDN")
+		}
+		if resp.StatusCode == http.StatusBadRequest {
+			return nil, errors.New("bad request")
+		}
+		return nil, fmt.Errorf("unexpected status code when loading persisted operation, statusCode: %d", resp.StatusCode)
+	}
+
+	var reader io.Reader = resp.Body
+
+	if resp.Header.Get("Content-Encoding") == "gzip" {
+		r, err := gzip.NewReader(resp.Body)
+		if err != nil {
+			return nil, errors.New("could not create gzip reader. " + err.Error())
+		}
+		defer r.Close()
+		reader = r
+	}
+
+	body, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, errors.New("could not read the response body. " + err.Error())
+	}
+
+	var po persistedoperation.PersistedOperation
+	err = json.Unmarshal(body, &po)
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(po.Body), nil
```

Comment:
If po.Body is never written to and only being read, we could reduce allocs by converting it unsafe to byte slice, otherwise keep it as is.